contracts/src/v0.8/vrf/dev/OptimismL1Fees.sol

```diff
@@ -0,0 +1,91 @@
+  /// @dev This option is available for the Coordinator and the Wrapper contract
+  uint8 internal constant L1_CALLDATA_GAS_COST_MODE = 1;
+  /// @dev Option 3: getL1FeeUpperBound() function from predeploy GasPriceOracle contract (available after Fjord upgrade)
+  /// @dev This option is available for the Coordinator and the Wrapper contract
+  uint8 internal constant L1_GAS_FEES_UPPER_BOUND_MODE = 2;
+
+  uint8 public s_L1FeeCalculationMode = L1_GAS_FEES_MODE;
+
+  /// @dev L1 fee coefficient can be applied to options 2 or 3 to reduce possibly inflated gas price
+  uint8 public s_L1FeeCoefficient = 100;
+
+  error InvalidL1FeeCalculationMode(uint8 mode);
+  error InvalidL1FeeCoefficient(uint8 coefficient);
+
+  event L1FeeCalculationSet(uint8 mode, uint8 coefficient);
+
+  function setL1FeeCalculation(uint8 mode, uint8 coefficient) external onlyOwner {
+    if (mode >= 3) {
+      revert InvalidL1FeeCalculationMode(mode);
+    }
+    if (coefficient == 0 || coefficient > 100) {
+      revert InvalidL1FeeCoefficient(coefficient);
+    }
+
+    s_L1FeeCalculationMode = mode;
+    s_L1FeeCoefficient = coefficient;
+
+    emit L1FeeCalculationSet(mode, coefficient);
+  }
+
+  function _getL1CostWeiForCalldata(bytes calldata data) internal view virtual returns (uint256) {
+    if (s_L1FeeCalculationMode == L1_GAS_FEES_MODE) {
+      return OVM_GASPRICEORACLE.getL1Fee(bytes.concat(data, L1_FEE_DATA_PADDING));
+    }
+    return _getL1CostWeiForCalldataSize(data.length);
+  }
+
+  function _getL1CostWeiForCalldataSize(uint256 calldataSizeBytes) internal view virtual returns (uint256) {
+    uint8 l1FeeCalculationMode = s_L1FeeCalculationMode;
+    if (l1FeeCalculationMode == L1_CALLDATA_GAS_COST_MODE) {
+      return (s_L1FeeCoefficient * _calculateOptimismL1DataFee(calldataSizeBytes)) / 100;
+    } else if (l1FeeCalculationMode == L1_GAS_FEES_UPPER_BOUND_MODE) {
+      // getL1FeeUpperBound expects unsigned fully RLP-encoded transaction size so we have to account for paddding bytes as well
+      return
+        (s_L1FeeCoefficient *
+          OVM_GASPRICEORACLE.getL1FeeUpperBound(calldataSizeBytes + L1_UNSIGNED_RLP_ENC_TX_DATA_BYTES_SIZE)) / 100;
+    }
+    revert InvalidL1FeeCalculationMode(l1FeeCalculationMode);
+  }
+
+  function _calculateOptimismL1DataFee(uint256 calldataSizeBytes) internal view returns (uint256) {
+    // reference: https://docs.optimism.io/stack/transactions/fees#ecotone
+    // also: https://github.com/ethereum-optimism/specs/blob/main/specs/protocol/exec-engine.md#ecotone-l1-cost-fee-changes-eip-4844-da
+    // we treat all bytes in the calldata payload as non-zero bytes (cost: 16 gas) because accurate estimation is too expensive
+    // we also have to account for the padding data and the signature data
+    uint256 l1GasUsed = (calldataSizeBytes + L1_TX_SIGNATURE_DATA_BYTES_SIZE + L1_UNSIGNED_RLP_ENC_TX_DATA_BYTES_SIZE) *
```

Comment:
I feel like we should hoist `L1_UNSIGNED_RLP_ENC_TX_DATA_BYTES_SIZE` up so that `_calculateOptimismL1DataFee` can maintain parity with `getL1FeeUpperBound`

Reply:
How many more bytes do you suggest in that case? I think we should probably test this first and measure the difference between `_calculateOptimismL1DataFee` and `getL1FeeUpperBound`. This padding size is still just the worst-case estimate.

Reply:
Not number of bytes but this.
```
if (l1FeeCalculationMode == L1_CALLDATA_GAS_COST_MODE) {
  return (s_L1FeeCoefficient * _calculateOptimismL1DataFee(calldataSizeBytes + L1_UNSIGNED_RLP_ENC_TX_DATA_BYTES_SIZE)) / 100;
}
```
function _calculateOptimismL1DataFee(uint256 calldataSizeBytes) internal view returns (uint256) {
    ...
    uint256 l1GasUsed = (calldataSizeBytes + L1_TX_SIGNATURE_DATA_BYTES_SIZE ~+ L1_UNSIGNED_RLP_ENC_TX_DATA_BYTES_SIZE~) * 16;

Reply:
Oh, I see. Yes, that makes sense 👍